### PR TITLE
Add additional filtering options

### PR DESCRIPTION
**Pull request by schustmi**
## Describe changes
This PR adds some additional filtering options that we want to use in the dashboard.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/zen_stores/sql_zen_store.py [hunk 1]
```diff
@@ -973,6 +973,7 @@ def filter_and_paginate(
             ValueError: if the filtered page number is out of bounds.
             RuntimeError: if the schema does not have a `to_model` method.
         """
+        query = query.distinct()
```
Comment (AlexejPenner) [inline src/zenml/zen_stores/sql_zen_store.py:976]: just for me to understand, why was this necessary only now?

**Review by AlexejPenner — COMMENTED** [review]
just two questions

File: src/zenml/models/v2/base/filter.py [hunk 2]
```diff
@@ -790,6 +790,67 @@ def _define_bool_filter(
...
+            # a ValueError. In this case, we already know that the filter
+            # will not produce any result and can simply ignore it.
+            pass
+
+        filter_ = self._define_filter(
+            column="name", value=value, operator=operator
+        )
+        conditions.append(filter_.generate_query_conditions(table=table))
+
+        return or_(False, *conditions)
```
[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/models/v2/base/filter.py:830]: Yes that's correct. Initially my method was more flexible and the conditions array could potentially be empty, which is why I added this safety `False` here so SQLAlchemy doesn't crash. I'll remove it

File: src/zenml/zen_stores/sql_zen_store.py (hunk 1 above)
Comment (schustmi) [reply src/zenml/zen_stores/sql_zen_store.py:976]: Tbh I don't really know. I encountered duplicate results once I allowed filtering tags by prefix (e.g. `tag="startswith:t"`), and if one entity contained multiple tags that matched this filter it appeared more than once in the result.

**Review by AlexejPenner — APPROVED** [review]
:seal:

**Comment by AlexejPenner** [comment]
@schustmi Anything in particular I should look at again?

**Comment by schustmi** [comment]
@AlexejPenner all of it ;)
I'm guessing the filtering pipelines by latest run status is the only major change that happened since your last review

File: src/zenml/models/v2/core/run_template.py [hunk 3]
```diff
@@ -390,4 +409,37 @@ def get_custom_filters(
...
+                == PipelineDeploymentSchema.id,
+                PipelineDeploymentSchema.pipeline_id == PipelineSchema.id,
+                self.generate_name_or_id_query_conditions(
+                    value=self.pipeline,
+                    table=PipelineSchema,
+                ),
+            )
+            custom_filters.append(pipeline_filter)
+
+        if self.stack:
```
Comment (AlexejPenner) [inline src/zenml/models/v2/core/run_template.py:433]:
```suggestion
        if self.stack is not None:
```

Shouldn't this be the same as the two above?

File: src/zenml/models/v2/core/pipeline_run.py [hunk 4]
```diff
@@ -682,6 +708,70 @@ def get_custom_filters(
...
+                CodeReferenceSchema.code_repository_id
+                == CodeRepositorySchema.id,
+                self.generate_name_or_id_query_conditions(
+                    value=self.code_repository,
+                    table=CodeRepositorySchema,
+                ),
+            )
+            custom_filters.append(code_repo_filter)
+
+        if self.model:
```
Comment (AlexejPenner) [inline src/zenml/models/v2/core/pipeline_run.py:754]: Any reason these are not also `is not None` - what am I missing??

**Review by AlexejPenner — APPROVED** [review]
Two small nits for consistency - else :seal:

File: src/zenml/models/v2/core/pipeline_run.py (hunk 4 above)
Comment (schustmi) [reply src/zenml/models/v2/core/pipeline_run.py:754]: Tbh i don't think it matters much for these, just empty strings would be treated differently. I guess for empty strings we don't actually want to include them in filters right? Every string in python starts/endswith/contains the empty string, which basically means the filter will always be true?

Comment (schustmi) [reply src/zenml/models/v2/core/pipeline_run.py:754]: I removed the `is not None` for all of them to disregard empty strings

**Review by AlexejPenner — APPROVED** [review]
:seal: - Happy to see filtering get better :)
